src/util/trace.h

```diff
@@ -11,39 +11,46 @@
 
 #ifdef ENABLE_TRACING
 
+// Setting SDT_USE_VARIADIC let's systemtap (sys/sdt.h) know that we want to use
+// the optional variadic macros to define tracepoints.
+#define SDT_USE_VARIADIC 1
+
+// Setting _SDT_HAS_SEMAPHORES let's systemtap (sys/sdt.h) know that we want to
+// use the optional semaphore feature for our tracepoints. This feature allows
+// us to check if something is attached to a tracepoint. We only want to prepare
+// some potentially expensive tracepoint arguments, if the tracepoint is being
+// used. Here, an expensive argument preparation could, for example, be
+// calculating a hash or serialization of a data structure.
+#define _SDT_HAS_SEMAPHORES 1
+
+// Used to define a counting semaphore for a tracepoint. This semaphore is
+// automatically incremented by tracing frameworks (bpftrace, bcc, libbpf, ...)
+// upon attaching to the tracepoint and decremented when detaching. This needs
+// to be a global variable. It's placed in the '.probes' ELF section.
+#define TRACEPOINT_SEMAPHORE(context, event) \
+    unsigned short context##_##event##_semaphore __attribute__((section(".probes")))
```

Comment:
I remember trying a few combinations of this and in all cases the compiler/liker failed to find the semaphore. I just retried with the following patch which I think is similar to what you describe.

```patch
diff --git a/src/util/trace.h b/src/util/trace.h
index a4f0ea1ca4..8d7f6bdb74 100644
--- a/src/util/trace.h
+++ b/src/util/trace.h
@@ -27,8 +27,8 @@
 // automatically incremented by tracing frameworks (bpftrace, bcc, libbpf, ...)
 // upon attaching to the tracepoint and decremented when detaching. This needs
 // to be a global variable. It's placed in the '.probes' ELF section.
-#define TRACEPOINT_SEMAPHORE(context, event) \
-    unsigned short context##_##event##_semaphore __attribute__((section(".probes")))
+#define TRACEPOINT_SEMAPHORE(context, event)
+//    unsigned short context##_##event##_semaphore __attribute__((section(".probes")))

 #include <sys/sdt.h>

@@ -39,6 +39,7 @@
 // tracepoint is active before preparing its arguments.
 #define TRACEPOINT(context, event, ...)                                         \
     do {                                                                        \
+        static unsigned short context##_##event##_semaphore __attribute__((section(".probes"))); \
         if (TRACEPOINT_ACTIVE(context, event)) {                                \
             STAP_PROBEV(context, event __VA_OPT__(, ) __VA_ARGS__);                           \
         }                                                                       \

```

This (1) breaks the extra [`if(TRACEPOINT_ACTIVE(context, event))`](https://github.com/bitcoin/bitcoin/pull/26593/files#diff-86c34c3a3e5946d9f4bd754a5bf9c970b6a44a72c0ba216f1a2457a479ee9651R50) and (2) the linker complains it can't find the semaphores (e.g. `undefined reference to `utxocache_spent_semaphore'`). 

Otherwise, if we find a way to do it, I'd be happy to do it as I'm not too excited about having the semaphores as globals..